felix/fv/pre_dnat_test.go

```diff
@@ -229,6 +246,168 @@ var _ = infrastructure.DatastoreDescribe("pre-dnat with initialized Felix, 2 wor
 						cc.CheckConnectivity()
 					})
+
+					It("increments Prometheus gauge proportionally to programming rules", func() {
+						dataplaneProgrammed := func() bool {
+							// Check if a particular known chain has been programmed into the mangle table.
+							mangleChains := tc.Felixes[0].IPTablesChains("mangle")
+							if _, ok := mangleChains["cali-fh-eth0"]; !ok {
+								return false
+							}
+
+							if len(mangleChains["cali-fh-eth0"]) == 0 {
+								return false
+							}
+
+							return true
+						}
+						Eventually(dataplaneProgrammed).Should(BeTrue(), "Dataplane never got fully programmed")
+
+						// A test-HEP to apply.
+						hep := api.NewHostEndpoint()
+						hep.Name = "t0"
+						hep.Spec.Node = tc.Felixes[0].Hostname
+						hep.Labels = map[string]string{"abc123": "true"}
+						hep.Spec.InterfaceName = "*"
+
+						// A test-GNP for the HEP.
+						policy := api.NewGlobalNetworkPolicy()
+						policy.Name = "allow-ingress-8055-1"
+						order := float64(11)
+						policy.Spec.Order = &order
+						policy.Spec.PreDNAT = true
+						policy.Spec.ApplyOnForward = true
+						protocol := numorstring.ProtocolFromString("tcp")
+						testPort := numorstring.SinglePort(9999)
+						ports := numorstring.SinglePort(8055)
+						metricsPort := numorstring.SinglePort(9091)
+						policy.Spec.Ingress = []api.Rule{{
+							Action:   api.Allow,
+							Protocol: &protocol,
+							Destination: api.EntityRule{Ports: []numorstring.Port{
+								testPort,
+								ports,
+								metricsPort,
+							}},
+						}}
+						policy.Spec.Selector = "has(abc123)"
+						// The same GNP but with stateful fields set.
+						var appliedGNP *api.GlobalNetworkPolicy
+
+						mangleRulesMetric := tc.Felixes[0].PromMetric("felix_iptables_rules{ip_version=\"4\",table=\"mangle\"}")
+						filterRulesMetric := tc.Felixes[0].PromMetric("felix_iptables_rules{ip_version=\"4\",table=\"filter\"}")
+						Eventually(mangleRulesMetric.Int, "5s").ShouldNot(BeZero(), "Metrics traffic was never allowed")
+
+						collectMetrics := func() (mangleMetric int, filterMetric int) {
+							mangleMetric, err := mangleRulesMetric.Int()
+							Expect(err).NotTo(HaveOccurred())
+
+							filterMetric, err = filterRulesMetric.Int()
+							Expect(err).NotTo(HaveOccurred())
+
+							return mangleMetric, filterMetric
+						}
+						// Perform database changes in steps.
+						type operation struct {
+							description string
+							do          func()
+						}
+						operations := []operation{
+							{"Creating a HEP and pre-DNAT GNP", func() {
+								var err error
+								curMangleRulesMetric, err := mangleRulesMetric.Int()
+								Expect(err).NotTo(HaveOccurred())
+
+								_, err = client.HostEndpoints().Create(utils.Ctx, hep, utils.NoOptions)
+								Expect(err).NotTo(HaveOccurred(), "Failed to create HEP")
+
+								appliedGNP, err = client.GlobalNetworkPolicies().Create(utils.Ctx, policy, utils.NoOptions)
```

Comment:
You should create the GNP first.  Felix won't program the GNP unless there's a HEP but it'll program the HEP without a GNP.  So, if you do HEP then GNP, you might see an intermediate step, whereas it'll be atomic the other way.